{
    "DWF": {
        "references": [
            "https://www.suse.com/security/cve/CVE-2021-29532/",
            "https://security.archlinux.org/CVE-2021-29532"
        ]
    },
    "MITRE": {
        "CVE_data_meta": {
            "ASSIGNER": "security-advisories@github.com",
            "ID": "CVE-2021-29532",
            "STATE": "PUBLIC",
            "TITLE": "Heap out of bounds read in `RaggedCross`"
        },
        "affects": {
            "vendor": {
                "vendor_data": [
                    {
                        "product": {
                            "product_data": [
                                {
                                    "product_name": "tensorflow",
                                    "version": {
                                        "version_data": [
                                            {
                                                "version_value": "< 2.1.4"
                                            },
                                            {
                                                "version_value": ">= 2.2.0, < 2.2.3"
                                            },
                                            {
                                                "version_value": ">= 2.3.0, < 2.3.3"
                                            },
                                            {
                                                "version_value": ">= 2.4.0, < 2.4.2"
                                            }
                                        ]
                                    }
                                }
                            ]
                        },
                        "vendor_name": "tensorflow"
                    }
                ]
            }
        },
        "data_format": "MITRE",
        "data_type": "CVE",
        "data_version": "4.0",
        "description": {
            "description_data": [
                {
                    "lang": "eng",
                    "value": "TensorFlow is an end-to-end open source platform for machine learning. An attacker can force accesses outside the bounds of heap allocated arrays by passing in invalid tensor values to `tf.raw_ops.RaggedCross`. This is because the implementation(https://github.com/tensorflow/tensorflow/blob/efea03b38fb8d3b81762237dc85e579cc5fc6e87/tensorflow/core/kernels/ragged_cross_op.cc#L456-L487) lacks validation for the user supplied arguments. Each of the above branches call a helper function after accessing array elements via a `*_list[next_*]` pattern, followed by incrementing the `next_*` index. However, as there is no validation that the `next_*` values are in the valid range for the corresponding `*_list` arrays, this results in heap OOB reads. The fix will be included in TensorFlow 2.5.0. We will also cherrypick this commit on TensorFlow 2.4.2, TensorFlow 2.3.3, TensorFlow 2.2.3 and TensorFlow 2.1.4, as these are also affected and still in supported range."
                }
            ]
        },
        "impact": {
            "cvss": {
                "attackComplexity": "HIGH",
                "attackVector": "LOCAL",
                "availabilityImpact": "LOW",
                "baseScore": 2.5,
                "baseSeverity": "LOW",
                "confidentialityImpact": "NONE",
                "integrityImpact": "NONE",
                "privilegesRequired": "LOW",
                "scope": "UNCHANGED",
                "userInteraction": "NONE",
                "vectorString": "CVSS:3.1/AV:L/AC:H/PR:L/UI:N/S:U/C:N/I:N/A:L",
                "version": "3.1"
            }
        },
        "problemtype": {
            "problemtype_data": [
                {
                    "description": [
                        {
                            "lang": "eng",
                            "value": "CWE-125: Out-of-bounds Read"
                        }
                    ]
                }
            ]
        },
        "references": {
            "reference_data": [
                {
                    "name": "https://github.com/tensorflow/tensorflow/security/advisories/GHSA-j47f-4232-hvv8",
                    "refsource": "CONFIRM",
                    "url": "https://github.com/tensorflow/tensorflow/security/advisories/GHSA-j47f-4232-hvv8"
                },
                {
                    "name": "https://github.com/tensorflow/tensorflow/commit/44b7f486c0143f68b56c34e2d01e146ee445134a",
                    "refsource": "MISC",
                    "url": "https://github.com/tensorflow/tensorflow/commit/44b7f486c0143f68b56c34e2d01e146ee445134a"
                }
            ]
        },
        "source": {
            "advisory": "GHSA-j47f-4232-hvv8",
            "discovery": "UNKNOWN"
        }
    },
    "NVD": {
        "configurations": {
            "CVE_data_version": "4.0",
            "nodes": [
                {
                    "children": [],
                    "cpe_match": [
                        {
                            "cpe23Uri": "cpe:2.3:a:google:tensorflow:*:*:*:*:*:*:*:*",
                            "cpe_name": [],
                            "versionEndExcluding": "2.1.4",
                            "vulnerable": true
                        },
                        {
                            "cpe23Uri": "cpe:2.3:a:google:tensorflow:*:*:*:*:*:*:*:*",
                            "cpe_name": [],
                            "versionEndExcluding": "2.2.3",
                            "versionStartIncluding": "2.2.0",
                            "vulnerable": true
                        },
                        {
                            "cpe23Uri": "cpe:2.3:a:google:tensorflow:*:*:*:*:*:*:*:*",
                            "cpe_name": [],
                            "versionEndExcluding": "2.3.3",
                            "versionStartIncluding": "2.3.0",
                            "vulnerable": true
                        },
                        {
                            "cpe23Uri": "cpe:2.3:a:google:tensorflow:*:*:*:*:*:*:*:*",
                            "cpe_name": [],
                            "versionEndExcluding": "2.4.2",
                            "versionStartIncluding": "2.4.0",
                            "vulnerable": true
                        }
                    ],
                    "operator": "OR"
                }
            ]
        },
        "cve": {
            "CVE_data_meta": {
                "ASSIGNER": "security-advisories@github.com",
                "ID": "CVE-2021-29532"
            },
            "data_format": "MITRE",
            "data_type": "CVE",
            "data_version": "4.0",
            "description": {
                "description_data": [
                    {
                        "lang": "en",
                        "value": "TensorFlow is an end-to-end open source platform for machine learning. An attacker can force accesses outside the bounds of heap allocated arrays by passing in invalid tensor values to `tf.raw_ops.RaggedCross`. This is because the implementation(https://github.com/tensorflow/tensorflow/blob/efea03b38fb8d3b81762237dc85e579cc5fc6e87/tensorflow/core/kernels/ragged_cross_op.cc#L456-L487) lacks validation for the user supplied arguments. Each of the above branches call a helper function after accessing array elements via a `*_list[next_*]` pattern, followed by incrementing the `next_*` index. However, as there is no validation that the `next_*` values are in the valid range for the corresponding `*_list` arrays, this results in heap OOB reads. The fix will be included in TensorFlow 2.5.0. We will also cherrypick this commit on TensorFlow 2.4.2, TensorFlow 2.3.3, TensorFlow 2.2.3 and TensorFlow 2.1.4, as these are also affected and still in supported range."
                    }
                ]
            },
            "problemtype": {
                "problemtype_data": [
                    {
                        "description": [
                            {
                                "lang": "en",
                                "value": "CWE-125"
                            }
                        ]
                    }
                ]
            },
            "references": {
                "reference_data": [
                    {
                        "name": "https://github.com/tensorflow/tensorflow/commit/44b7f486c0143f68b56c34e2d01e146ee445134a",
                        "refsource": "MISC",
                        "tags": [
                            "Patch",
                            "Third Party Advisory"
                        ],
                        "url": "https://github.com/tensorflow/tensorflow/commit/44b7f486c0143f68b56c34e2d01e146ee445134a"
                    },
                    {
                        "name": "https://github.com/tensorflow/tensorflow/security/advisories/GHSA-j47f-4232-hvv8",
                        "refsource": "CONFIRM",
                        "tags": [
                            "Exploit",
                            "Patch",
                            "Third Party Advisory"
                        ],
                        "url": "https://github.com/tensorflow/tensorflow/security/advisories/GHSA-j47f-4232-hvv8"
                    }
                ]
            }
        },
        "impact": {
            "baseMetricV2": {
                "acInsufInfo": false,
                "cvssV2": {
                    "accessComplexity": "LOW",
                    "accessVector": "LOCAL",
                    "authentication": "NONE",
                    "availabilityImpact": "PARTIAL",
                    "baseScore": 3.6,
                    "confidentialityImpact": "PARTIAL",
                    "integrityImpact": "NONE",
                    "vectorString": "AV:L/AC:L/Au:N/C:P/I:N/A:P",
                    "version": "2.0"
                },
                "exploitabilityScore": 3.9,
                "impactScore": 4.9,
                "obtainAllPrivilege": false,
                "obtainOtherPrivilege": false,
                "obtainUserPrivilege": false,
                "severity": "LOW",
                "userInteractionRequired": false
            },
            "baseMetricV3": {
                "cvssV3": {
                    "attackComplexity": "LOW",
                    "attackVector": "LOCAL",
                    "availabilityImpact": "HIGH",
                    "baseScore": 7.1,
                    "baseSeverity": "HIGH",
                    "confidentialityImpact": "HIGH",
                    "integrityImpact": "NONE",
                    "privilegesRequired": "LOW",
                    "scope": "UNCHANGED",
                    "userInteraction": "NONE",
                    "vectorString": "CVSS:3.1/AV:L/AC:L/PR:L/UI:N/S:U/C:H/I:N/A:H",
                    "version": "3.1"
                },
                "exploitabilityScore": 1.8,
                "impactScore": 5.2
            }
        },
        "lastModifiedDate": "2021-07-26T15:56Z",
        "publishedDate": "2021-05-14T20:15Z"
    }
}